File: VB/RefreshHelperClass.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections
Imports DevExpress.XtraGrid
Imports DevExpress.Utils
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Grid

Namespace DevExpress.XtraGrid.Helpers

	Public Class GridControlState

		Public Structure RowInfo
			Public Id As Object
			Public level As Integer
		End Structure

		Public Structure ViewDescriptor
			Public relationName As String
			Public keyFieldName As String

			Public Sub New(ByVal relationName As String, ByVal keyFieldName As String)
				Me.relationName = relationName
				Me.keyFieldName = keyFieldName
			End Sub
		End Structure

		Public Class ViewState
			Private gridState As GridControlState
			Private parent As ViewState
			Private descriptor As ViewDescriptor
			Private saveExpList_Renamed As ArrayList
			Private saveSelList_Renamed As ArrayList
			Private saveMasterRowsList_Renamed As ArrayList
			Private visibleRowIndex As Integer = -1
			Private detailViews_Renamed As Hashtable
			Private horzScrollPos As Integer
			Private cellSelection_Renamed As Hashtable

			Protected Sub New(ByVal gridState As GridControlState, ByVal descriptor As ViewDescriptor)
				Me.gridState = gridState
				Me.descriptor = descriptor
				Me.parent = Nothing
			End Sub

			Protected Sub New(ByVal parent As ViewState, ByVal descriptor As ViewDescriptor)
				Me.parent = parent
				Me.gridState = parent.gridState
				Me.descriptor = descriptor
			End Sub

			Public Shared Function Create(ByVal gridState As GridControlState, ByVal view As GridView) As ViewState
				If (Not gridState.viewDescriptors.ContainsKey(view.LevelName)) Then
					Return Nothing
				End If
				Dim state As New ViewState(gridState, CType(gridState.viewDescriptors(view.LevelName), ViewDescriptor))
				Return state
			End Function

			Private Shared Function Create(ByVal parent As ViewState, ByVal view As GridView) As ViewState
				If (Not parent.gridState.viewDescriptors.ContainsKey(view.LevelName)) Then
					Return Nothing
				End If
				Dim state As New ViewState(parent, CType(parent.gridState.viewDescriptors(view.LevelName), ViewDescriptor))
				Return state
			End Function

			Public Function IsLevel(ByVal level As String) As Boolean
				Return level = Me.descriptor.relationName
			End Function

			Public ReadOnly Property SaveExpList() As ArrayList
				Get
					If saveExpList_Renamed Is Nothing Then
						saveExpList_Renamed = New ArrayList()
					End If
					Return saveExpList_Renamed
				End Get
			End Property

			Public ReadOnly Property SaveSelList() As ArrayList
				Get
					If saveSelList_Renamed Is Nothing Then
						saveSelList_Renamed = New ArrayList()
					End If
					Return saveSelList_Renamed
				End Get
			End Property

			Public ReadOnly Property SaveMasterRowsList() As ArrayList
				Get
					If saveMasterRowsList_Renamed Is Nothing Then
						saveMasterRowsList_Renamed = New ArrayList()
					End If
					Return saveMasterRowsList_Renamed
				End Get
			End Property

			Public ReadOnly Property DetailViews() As Hashtable
				Get
					If detailViews_Renamed Is Nothing Then
						detailViews_Renamed = New Hashtable()
					End If
					Return detailViews_Renamed
				End Get
			End Property

			Public ReadOnly Property CellSelection() As Hashtable
				Get
					If cellSelection_Renamed Is Nothing Then
						cellSelection_Renamed = New Hashtable()
					End If
					Return cellSelection_Renamed
				End Get
			End Property

			Protected Function FindParentRowHandle(ByVal view As GridView, ByVal rowInfo As RowInfo, ByVal rowHandle As Integer) As Integer
				Dim result As Integer = view.GetParentRowHandle(rowHandle)
				Do While view.GetRowLevel(result) <> rowInfo.level
					result = view.GetParentRowHandle(result)
				Loop
				Return result
			End Function

			Protected Sub ExpandRowByRowInfo(ByVal view As GridView, ByVal rowInfo As RowInfo)
				Dim dataRowHandle As Integer = LocateRowByKeyValue(view, rowInfo.Id)
				If dataRowHandle <> GridControl.InvalidRowHandle Then
					Dim parentRowHandle As Integer = FindParentRowHandle(view, rowInfo, dataRowHandle)
					view.SetRowExpanded(parentRowHandle, True, False)
				End If
			End Sub

			Protected Function GetRowHandleToSelect(ByVal view As GridView, ByVal rowInfo As RowInfo) As Integer
				Dim dataRowHandle As Integer = LocateRowByKeyValue(view, rowInfo.Id)
				If dataRowHandle <> GridControl.InvalidRowHandle Then
					If view.GetRowLevel(dataRowHandle) <> rowInfo.level Then
						Return FindParentRowHandle(view, rowInfo, dataRowHandle)
					End If
				End If
				Return dataRowHandle
			End Function

			Protected Sub SelectRowByRowInfo(ByVal view As GridView, ByVal rowInfo As RowInfo, ByVal isFocused As Boolean)
				Dim rowHandle As Integer = GetRowHandleToSelect(view, rowInfo)
				If isFocused Then
					view.FocusedRowHandle = rowHandle
				Else
					If view.OptionsSelection.MultiSelectMode = GridMultiSelectMode.CellSelect Then
						Dim names() As String = TryCast(CellSelection(rowInfo.Id), String())
						If names IsNot Nothing Then
							For j As Integer = 0 To names.Length - 1
								view.SelectCell(rowHandle, view.Columns(names(j)))
							Next j
						End If
					Else
						view.SelectRow(rowHandle)
					End If
				End If
			End Sub

			Public Sub SaveSelectionViewInfo(ByVal view As GridView)
				SaveSelList.Clear()
				CellSelection.Clear()
				Dim selectionArray() As Integer = view.GetSelectedRows()
				If selectionArray IsNot Nothing Then ' otherwise we have a single focused but not selected row
					For i As Integer = 0 To selectionArray.Length - 1
						Dim dataRowHandle As Integer = AddRowToSelection(view, selectionArray(i))
						If view.OptionsSelection.MultiSelectMode = GridMultiSelectMode.CellSelect Then
							Dim columns() As GridColumn = view.GetSelectedCells(dataRowHandle)
							Dim names(columns.Length - 1) As String
							For j As Integer = 0 To columns.Length - 1
								names(j) = columns(j).FieldName
							Next j
							CellSelection(view.GetRowCellValue(dataRowHandle, descriptor.keyFieldName)) = names
						End If
					Next i
				End If
				AddRowToSelection(view, view.FocusedRowHandle)
			End Sub
			Private Function AddRowToSelection(ByVal view As GridView, ByVal handle As Integer) As Integer
				Dim rowInfo As RowInfo
				rowInfo.level = view.GetRowLevel(handle)
				If handle < 0 Then ' group row
					handle = view.GetDataRowHandleByGroupRowHandle(handle)
				End If
				rowInfo.Id = view.GetRowCellValue(handle, descriptor.keyFieldName)
				SaveSelList.Add(rowInfo)
				Return handle
			End Function

			Public Sub SaveExpansionViewInfo(ByVal view As GridView)
				If view.GroupedColumns.Count = 0 Then
					Return
				End If
				SaveExpList.Clear()
				For i As Integer = -1 To Integer.MinValue + 1 Step -1
					If (Not view.IsValidRowHandle(i)) Then
						Exit For
					End If
					If view.GetRowExpanded(i) Then
						Dim rowInfo As RowInfo
						Dim dataRowHandle As Integer = view.GetDataRowHandleByGroupRowHandle(i)
						rowInfo.Id = view.GetRowCellValue(dataRowHandle, descriptor.keyFieldName)
						rowInfo.level = view.GetRowLevel(i)
						SaveExpList.Add(rowInfo)
					End If
				Next i
			End Sub

			Public Sub SaveExpandedMasterRows(ByVal view As GridView)
				If view.GridControl.Views.Count = 1 Then
					Return
				End If
				SaveMasterRowsList.Clear()
				For i As Integer = 0 To view.DataRowCount - 1
					If view.GetMasterRowExpanded(i) Then
						Dim key As Object = view.GetRowCellValue(i, descriptor.keyFieldName)
						SaveMasterRowsList.Add(key)
						Dim detail As GridView = TryCast(view.GetVisibleDetailView(i), GridView)
						If detail IsNot Nothing Then
							Dim state As ViewState = ViewState.Create(Me, detail)
							If state IsNot Nothing Then
								DetailViews.Add(key, state)
								state.SaveState(detail)
							End If
						End If
					End If
				Next i
			End Sub

			Public Sub SaveVisibleIndex(ByVal view As GridView)
				visibleRowIndex = view.GetVisibleIndex(view.FocusedRowHandle) - view.TopRowIndex
			End Sub

			Public Sub LoadVisibleIndex(ByVal view As GridView)
				view.MakeRowVisible(view.FocusedRowHandle, True)
				view.TopRowIndex = view.GetVisibleIndex(view.FocusedRowHandle) - visibleRowIndex
			End Sub

			Public Sub LoadSelectionViewInfo(ByVal view As GridView)
				view.BeginSelection()
				Try
					view.ClearSelection()
					For i As Integer = 0 To SaveSelList.Count - 1
						SelectRowByRowInfo(view, CType(SaveSelList(i), RowInfo), i = SaveSelList.Count - 1)
					Next i
				Finally
					view.EndSelection()
				End Try
			End Sub

			Public Sub LoadExpansionViewInfo(ByVal view As GridView)
				If view.GroupedColumns.Count = 0 Then
					Return
				End If
				view.BeginUpdate()
				Try
					view.CollapseAllGroups()
					For Each info As RowInfo In SaveExpList
						ExpandRowByRowInfo(view, info)
					Next info
				Finally
					view.EndUpdate()
				End Try
			End Sub

			Public Sub LoadExpandedMasterRows(ByVal view As GridView)
				'view.BeginUpdate();
				Try
					view.CollapseAllDetails()
					For i As Integer = 0 To SaveMasterRowsList.Count - 1
						Dim rowHandle As Integer = LocateRowByKeyValue(view, SaveMasterRowsList(i))
						Dim state As ViewState = CType(DetailViews(SaveMasterRowsList(i)), ViewState)
						If state Is Nothing Then
							view.SetMasterRowExpanded(rowHandle, True)
						Else
							view.SetMasterRowExpandedEx(rowHandle, view.GetRelationIndex(rowHandle, state.descriptor.relationName), True)
							Dim detail As GridView = TryCast(view.GetVisibleDetailView(rowHandle), GridView)
							If detail IsNot Nothing Then
								state.LoadState(detail)
							End If
						End If
					Next i
				Finally
					'view.EndUpdate();
				End Try
			End Sub

			Private Function LocateRowByKeyValue(ByVal view As GridView, ByVal value As Object) As Integer
				If view.IsServerMode Then
					Return view.DataController.FindRowByValue(descriptor.keyFieldName, value)
				End If
				For i As Integer = 0 To view.DataRowCount - 1
					If Equals(value, view.GetRowCellValue(i, descriptor.keyFieldName)) Then
						Return i
					End If
				Next i
				Return GridControl.InvalidRowHandle
			End Function

			Public Sub SaveState(ByVal view As GridView)
				If ReferenceEquals(view.GridControl.FocusedView, view) Then
					gridState.focused = Me
				End If
				SaveExpandedMasterRows(view)
				SaveExpansionViewInfo(view)
				SaveSelectionViewInfo(view)
				SaveVisibleIndex(view)
				horzScrollPos = view.LeftCoord
			End Sub

			Public Sub LoadState(ByVal view As GridView)
				LoadExpandedMasterRows(view)
				LoadExpansionViewInfo(view)
				LoadSelectionViewInfo(view)
				LoadVisibleIndex(view)
				view.LeftCoord = horzScrollPos
				If ReferenceEquals(gridState.focused, Me) Then
					view.GridControl.FocusedView = view
				End If
			End Sub
		End Class

		Private viewDescriptors As Hashtable
		Private root As ViewState
		Private focused As ViewState

		Public Sub New(ByVal views() As ViewDescriptor)
			viewDescriptors = New Hashtable()
			For Each desc As ViewDescriptor In views
				viewDescriptors.Add(desc.relationName, desc)
			Next desc
		End Sub

		Public Sub SaveViewInfo(ByVal view As GridView)
			root = ViewState.Create(Me, view)
			root.SaveState(view)
		End Sub

		Public Sub LoadViewInfo(ByVal view As GridView)
			If root Is Nothing OrElse (Not root.IsLevel(view.LevelName)) Then
				Return
			End If
			root.LoadState(view)
		End Sub

	End Class
End Namespace

File: CS/RefreshHelperClass.cs
using System;
using System.Collections;
using DevExpress.XtraGrid;
using DevExpress.Utils;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Grid;

namespace DevExpress.XtraGrid.Helpers {

    public class GridControlState {

        public struct RowInfo {
            public object Id;
            public int level;
        }

        public struct ViewDescriptor {
            public string relationName;
            public string keyFieldName;

            public ViewDescriptor(string relationName, string keyFieldName) {
                this.relationName = relationName;
                this.keyFieldName = keyFieldName;
            }
        }

        public class ViewState {
            private GridControlState gridState;
            private ViewState parent;
            private ViewDescriptor descriptor;
            private ArrayList saveExpList;
            private ArrayList saveSelList;
            private ArrayList saveMasterRowsList;
            private int visibleRowIndex = -1;
            private Hashtable detailViews;
            private int horzScrollPos;
            private Hashtable cellSelection;

            protected ViewState(GridControlState gridState, ViewDescriptor descriptor) {
                this.gridState = gridState;
                this.descriptor = descriptor;
                this.parent = null;
            }

            protected ViewState(ViewState parent, ViewDescriptor descriptor) {
                this.parent = parent;
                this.gridState = parent.gridState;
                this.descriptor = descriptor;
            }

            public static ViewState Create(GridControlState gridState, GridView view) {
                if (!gridState.viewDescriptors.ContainsKey(view.LevelName)) return null;
                ViewState state = new ViewState(gridState, (ViewDescriptor)gridState.viewDescriptors[view.LevelName]);
                return state;
            }

            private static ViewState Create(ViewState parent, GridView view) {
                if (!parent.gridState.viewDescriptors.ContainsKey(view.LevelName)) return null;
                ViewState state = new ViewState(parent, (ViewDescriptor)parent.gridState.viewDescriptors[view.LevelName]);
                return state;
            }

            public bool IsLevel(string level) {
                return level == this.descriptor.relationName;
            }

            public ArrayList SaveExpList {
                get {
                    if (saveExpList == null)
                        saveExpList = new ArrayList();
                    return saveExpList;
                }
            }

            public ArrayList SaveSelList {
                get {
                    if (saveSelList == null)
                        saveSelList = new ArrayList();
                    return saveSelList;
                }
            }

            public ArrayList SaveMasterRowsList {
                get {
                    if (saveMasterRowsList == null)
                        saveMasterRowsList = new ArrayList();
                    return saveMasterRowsList;
                }
            }

            public Hashtable DetailViews {
                get {
                    if (detailViews == null)
                        detailViews = new Hashtable();
                    return detailViews;
                }
            }

            public Hashtable CellSelection {
                get {
                    if (cellSelection == null)
                        cellSelection = new Hashtable();
                    return cellSelection;
                }
            }

            protected int FindParentRowHandle(GridView view, RowInfo rowInfo, int rowHandle) {
                int result = view.GetParentRowHandle(rowHandle);
                while (view.GetRowLevel(result) != rowInfo.level)
                    result = view.GetParentRowHandle(result);
                return result;
            }

            protected void ExpandRowByRowInfo(GridView view, RowInfo rowInfo) {
                int dataRowHandle = LocateRowByKeyValue(view, rowInfo.Id);
                if (dataRowHandle != GridControl.InvalidRowHandle) {
                    int parentRowHandle = FindParentRowHandle(view, rowInfo, dataRowHandle);
                    view.SetRowExpanded(parentRowHandle, true, false);
                }
            }

            protected int GetRowHandleToSelect(GridView view, RowInfo rowInfo) {
                int dataRowHandle = LocateRowByKeyValue(view, rowInfo.Id);
                if (dataRowHandle != GridControl.InvalidRowHandle)
                    if (view.GetRowLevel(dataRowHandle) != rowInfo.level)
                        return FindParentRowHandle(view, rowInfo, dataRowHandle);
                return dataRowHandle;
            }

            protected void SelectRowByRowInfo(GridView view, RowInfo rowInfo, bool isFocused) {
                int rowHandle = GetRowHandleToSelect(view, rowInfo);
                if (isFocused)
                    view.FocusedRowHandle = rowHandle;
                else {
                    if (view.OptionsSelection.MultiSelectMode == GridMultiSelectMode.CellSelect) {
                        string[] names = CellSelection[rowInfo.Id] as string[];
                        if (names != null) {
                            for (int j = 0; j < names.Length; j++) {
                                view.SelectCell(rowHandle, view.Columns[names[j]]);
                            }
                        }
                    }
                    else {
                        view.SelectRow(rowHandle);
                    }
                }
            }

            public void SaveSelectionViewInfo(GridView view) {
                SaveSelList.Clear();
                CellSelection.Clear();
                int[] selectionArray = view.GetSelectedRows();
                if (selectionArray != null)  // otherwise we have a single focused but not selected row
                    for (int i = 0; i < selectionArray.Length; i++) {
                        int dataRowHandle = AddRowToSelection(view, selectionArray[i]);
                        if (view.OptionsSelection.MultiSelectMode == GridMultiSelectMode.CellSelect) {
                            GridColumn[] columns = view.GetSelectedCells(dataRowHandle);
                            string[] names = new string[columns.Length];
                            for (int j = 0; j < columns.Length; j++) {
                                names[j] = columns[j].FieldName;
                            }
                            CellSelection[view.GetRowCellValue(dataRowHandle, descriptor.keyFieldName)] = names;
                        }
                    }
                AddRowToSelection(view, view.FocusedRowHandle);
            }
            private int AddRowToSelection(GridView view, int handle) {
                RowInfo rowInfo;
                rowInfo.level = view.GetRowLevel(handle);
                if (handle < 0) // group row
                    handle = view.GetDataRowHandleByGroupRowHandle(handle);
                rowInfo.Id = view.GetRowCellValue(handle, descriptor.keyFieldName);
                SaveSelList.Add(rowInfo);
                return handle;
            }

            public void SaveExpansionViewInfo(GridView view) {
                if (view.GroupedColumns.Count == 0) return;
                SaveExpList.Clear();
                for (int i = -1; i > int.MinValue; i--) {
                    if (!view.IsValidRowHandle(i)) break;
                    if (view.GetRowExpanded(i)) {
                        RowInfo rowInfo;
                        int dataRowHandle = view.GetDataRowHandleByGroupRowHandle(i);
                        rowInfo.Id = view.GetRowCellValue(dataRowHandle, descriptor.keyFieldName);
                        rowInfo.level = view.GetRowLevel(i);
                        SaveExpList.Add(rowInfo);
                    }
                }
            }

            public void SaveExpandedMasterRows(GridView view) {
                if (view.GridControl.Views.Count == 1) return;
                SaveMasterRowsList.Clear();
                for (int i = 0; i < view.DataRowCount; i++)
                    if (view.GetMasterRowExpanded(i)) {
                        object key = view.GetRowCellValue(i, descriptor.keyFieldName);
                        SaveMasterRowsList.Add(key);
                        GridView detail = view.GetVisibleDetailView(i) as GridView;
                        if (detail != null) {
                            ViewState state = ViewState.Create(this, detail);
                            if (state != null) {
                                DetailViews.Add(key, state);
                                state.SaveState(detail);
                            }
                        }
                    }
            }

            public void SaveVisibleIndex(GridView view) {
                visibleRowIndex = view.GetVisibleIndex(view.FocusedRowHandle) - view.TopRowIndex;
            }

            public void LoadVisibleIndex(GridView view) {
                view.MakeRowVisible(view.FocusedRowHandle, true);
                view.TopRowIndex = view.GetVisibleIndex(view.FocusedRowHandle) - visibleRowIndex;
            }

            public void LoadSelectionViewInfo(GridView view) {
                view.DataController.EnsureFindRowByValueCache(view.DataController.Columns[descriptor.keyFieldName], view.RowCount);
                view.BeginSelection();
                try {
                    view.ClearSelection();
                    for (int i = 0; i < SaveSelList.Count; i++)
                        SelectRowByRowInfo(view, (RowInfo)SaveSelList[i], i == SaveSelList.Count - 1);
                }
                finally {
                    view.EndSelection();
                    view.DataController.DestroyFindRowByValueCache();
                }
            }

            public void LoadExpansionViewInfo(GridView view) {
                if (view.GroupedColumns.Count == 0) return;
                view.BeginUpdate();
                try {
                    view.CollapseAllGroups();
                    foreach (RowInfo info in SaveExpList)
                        ExpandRowByRowInfo(view, info);
                }
                finally {
                    view.EndUpdate();
                }
            }

            public void LoadExpandedMasterRows(GridView view) {
                //view.BeginUpdate();
                try {
                    view.CollapseAllDetails();
                    for (int i = 0; i < SaveMasterRowsList.Count; i++) {
                        int rowHandle = LocateRowByKeyValue(view, SaveMasterRowsList[i]);
                        ViewState state = (ViewState)DetailViews[SaveMasterRowsList[i]];
                        if (state == null) {
                            view.SetMasterRowExpanded(rowHandle, true);
                        }
                        else {
                            view.SetMasterRowExpandedEx(rowHandle, view.GetRelationIndex(rowHandle, state.descriptor.relationName), true);
                            GridView detail = view.GetVisibleDetailView(rowHandle) as GridView;
                            if (detail != null) {
                                state.LoadState(detail);
                            }
                        }
                    }
                }
                finally {
                    //view.EndUpdate();
                }
            }

            private int LocateRowByKeyValue(GridView view, object value) {
                if (view.IsServerMode) {
                    return view.DataController.FindRowByValue(descriptor.keyFieldName, value);
                }
                for (int i = 0; i < view.DataRowCount; i++) {
                    if (Equals(value, view.GetRowCellValue(i, descriptor.keyFieldName)))
                        return i;
                }
                return GridControl.InvalidRowHandle;
            }

            public void SaveState(GridView view) {
                if (ReferenceEquals(view.GridControl.FocusedView, view)) gridState.focused = this;
                SaveExpandedMasterRows(view);
                SaveExpansionViewInfo(view);
                SaveSelectionViewInfo(view);
                SaveVisibleIndex(view);
                horzScrollPos = view.LeftCoord;
            }

            public void LoadState(GridView view) {
                LoadExpandedMasterRows(view);
                LoadExpansionViewInfo(view);
                LoadSelectionViewInfo(view);
                LoadVisibleIndex(view);
                view.LeftCoord = horzScrollPos;
                if (ReferenceEquals(gridState.focused, this)) view.GridControl.FocusedView = view;
            }
        }

        private Hashtable viewDescriptors;
        private ViewState root;
        private ViewState focused;

        public GridControlState(ViewDescriptor[] views) {
            viewDescriptors = new Hashtable();
            foreach (ViewDescriptor desc in views) {
                viewDescriptors.Add(desc.relationName, desc);
            }
        }

        public void SaveViewInfo(GridView view) {
            root = ViewState.Create(this, view);
            root.SaveState(view);
        }

        public void LoadViewInfo(GridView view) {
            if (root == null || !root.IsLevel(view.LevelName)) return;
            root.LoadState(view);
        }

    }
}